File: implementations/go/harness/rule.go
package main

// Kind identifies whether a Finding is a pass, a failure, or a "not applicable" skip.
type Kind int

const (
	Pass Kind = iota
	Fail
	Skip
)

// Finding is a single reported outcome within a rule's section output.
// For Skip, Name carries the skip message and Reason is unused.
type Finding struct {
	Kind   Kind
	Name   string
	Reason string
}

// RuleResult is everything one rule check contributes: its section header
// and the ordered list of findings to print under it.
type RuleResult struct {
	Section  string
	Findings []Finding
}

func passFinding(name string) Finding {
	return Finding{Kind: Pass, Name: name}
}

func failFinding(name, reason string) Finding {
	return Finding{Kind: Fail, Name: name, Reason: reason}
}

func skipFinding(message string) Finding {
	return Finding{Kind: Skip, Name: message}
}

File: implementations/kotlin-springboot/harness/src/Rule.kt
package harness

/** The kind of a Finding — pass, fail, or a skip meaning "not applicable". */
enum class Kind { PASS, FAIL, SKIP }

/** A single item within one rule's section output. For a Skip, name holds the skip message. */
data class Finding(val kind: Kind, val name: String, val reason: String = "")

/** The result returned by a single rule — a section header + the list of Findings to print under it. */
class RuleResult(val section: String) {
    val findings: MutableList<Finding> = mutableListOf()

    fun add(finding: Finding) {
        findings.add(finding)
    }

    fun count(kind: Kind): Int = findings.count { it.kind == kind }
}

fun passFinding(name: String): Finding = Finding(Kind.PASS, name)
fun failFinding(name: String, reason: String): Finding = Finding(Kind.FAIL, name, reason)
fun skipFinding(message: String): Finding = Finding(Kind.SKIP, message)

/** The rule function signature — takes the projectRoot absolute path and returns a RuleResult. */
typealias Rule = (String) -> RuleResult
